ibc-clients/cw-context/src/context/client_ctx.rs

```diff
@@ -0,0 +1,179 @@
+//! Implementation of the `ClientValidationContext` and `ClientExecutionContext`
+//! traits for the `Context` type.
+use ibc_client_wasm_types::client_state::ClientState as WasmClientState;
+use ibc_client_wasm_types::consensus_state::ConsensusState as WasmConsensusState;
+use ibc_core::client::context::{ClientExecutionContext, ClientValidationContext};
+use ibc_core::client::types::error::ClientError;
+use ibc_core::client::types::Height;
+use ibc_core::handler::types::error::ContextError;
+use ibc_core::host::types::identifiers::ClientId;
+use ibc_core::host::types::path::{iteration_key, ClientConsensusStatePath, ClientStatePath};
+use ibc_core::primitives::proto::{Any, Protobuf};
+use ibc_core::primitives::Timestamp;
+
+use super::Context;
+use crate::api::ClientType;
+use crate::utils::AnyCodec;
+
+impl<'a, C: ClientType<'a>> ClientValidationContext for Context<'a, C> {
+    type ClientStateRef = C::ClientState;
+    type ConsensusStateRef = C::ConsensusState;
+
+    fn client_state(&self, _client_id: &ClientId) -> Result<Self::ClientStateRef, ContextError> {
+        let client_state_value = self.retrieve(ClientStatePath::leaf())?;
+
+        let any_wasm: WasmClientState = Protobuf::<Any>::decode(client_state_value.as_slice())
+            .map_err(|e| ClientError::Other {
+                description: e.to_string(),
+            })?;
+
+        let sov_client_state = C::ClientState::decode_thru_any(any_wasm.data)?;
+
+        Ok(sov_client_state)
+    }
+
+    fn consensus_state(
+        &self,
+        client_cons_state_path: &ClientConsensusStatePath,
+    ) -> Result<Self::ConsensusStateRef, ContextError> {
+        let consensus_state_value = self.retrieve(client_cons_state_path.leaf())?;
+
+        let any_wasm: WasmConsensusState =
+            C::ConsensusState::decode_thru_any(consensus_state_value)?;
+
+        let consensus_state = C::ConsensusState::decode_thru_any(any_wasm.data)?;
+
+        Ok(consensus_state)
+    }
+
+    fn client_update_meta(
+        &self,
+        _client_id: &ClientId,
+        height: &Height,
+    ) -> Result<(Timestamp, Height), ContextError> {
+        let time_key = self.client_update_time_key(height);
+
+        let time_vec = self.retrieve(time_key)?;
+
+        let time = u64::from_be_bytes(time_vec.try_into().expect("invalid timestamp"));
+
+        let timestamp =
+            Timestamp::from_nanoseconds(time).map_err(ClientError::InvalidPacketTimestamp)?;
+
+        let height_key = self.client_update_height_key(height);
+
+        let revision_height_vec = self.retrieve(height_key)?;
+
+        let revision_height =
+            u64::from_be_bytes(revision_height_vec.try_into().expect("invalid height"));
```

Comment:
calling `.expect` here. maybe propagate the `Err`.

Reply:
[6de21ed](https://github.com/cosmos/ibc-rs/pull/1167/commits/6de21edb62137b9d74daf1bcacf3b626e45013cc)